### ISSUE TITLE
Reusing permissions

### ISSUE THREAD

**Issue by altschuler** [hasura/graphql-engine#6400]
It would be very useful to be able to reuse permission definitions in some way. Let's say you have these tables (should be self-explanatory, a basic setup for groups with users who can author comment on posts)

```
user
group
post
comment

group_member (user+group relation)
group_post (post+group relation)
group_post_comment (post+comment relation)
```

The overarching permission setup here is through the `group_member` table. If there is an entry for a given group and a given user, that user can access any post and comment in the group.

Currently I have to duplicate this permission check ("user is member of group") in the `group`, `post`, `comment`, `group_post` and `group_post_comment`. All that duplication makes the configuration tedious and error-prone.

**What I'm suggesting is _some_ sort of functionality to indicate that a user has access to a `post` if they have access to the corresponding `group`, ie. delegate a select/insert/update/delete permission to another type.**

Is there currently an easier way to handle this than the duplication permissions? I realize this is a huge task, and would likely require a remake of the permission editor, but I'd like to know if anyone else is struggling a bit with this.

All that said, it's not impossibly hard to do this currently, it could just be very much easier.

(Sorry if this has come up before, I couldn't find any issues mentioning it)

**Comment by tirumaraiselvan** [comment]
This is a very valid concern. Another way of reusing permissions is to allow inheritance of roles: https://github.com/hasura/graphql-engine/issues/611 so for e.g. you can define a "group" role =  group's permissions (manually defined) + post's permissions (inherited). I am guessing this may not be what you are looking for since it _appears_ you want to do inheritance based on some sort of arbitrary mappings. This would definitely require little bit more framing since it is very vague atm.

Having said that, inherited roles is something that we are working on.

**Comment by altschuler** [comment]
Inherited roles sounds great, but unfortunately wont help in this case because as you said it's based on arbitrary mappings. It's more about reusing snippets of expressions and/or delegating it to another type:

1) Delegation by foreign key: on a `group_post` (a table containing posts in a group), with a foreign key to `groups`, you would say that if you have access to the referenced group, you also have access to the group_post.
2) "Snippet" reuse: somewhere in the console UI on each table, you would define named custom permission expressions. Eg. on the `groups` table, you'd define an expression called "is_member" as something like `{"memberships": { "user_id": { "_eq": "x-hasura-user-id" } } }`. Anywhere that you can build expressions for a `groups` table you would then be able to select `is_member` from the list of fields to configure (in addition to all the normally defined fields like `name`, `id`, etc). Eg in a `group_posts` permission maybe it would look like `{ "group": { "is_member": true } }`. If you can parameterize it somehow it would be even more flexible.

I think the latter would be the more generic option and would cover the same use cases as delegation by fkey, although in a slightly more complex way.

I hope that makes sense :)

**Comment by esseswann** [comment]
>1. Delegation by foreign key: on a group_post (a table containing posts in a group), with a foreign key to groups, you would say that if you have access to the referenced group, you also have access to the group_post.

This is probably the a valid approach since if the referenced table row's key is not accessible because of some permissions seems logical that another table rows referencing it through foreign key shouldn't also be accessible.
The problem is a possible explosion of joins when the algorithm is gonna traverse upwards and for each table does the same thing -- if the rows having foregin keys are accessible through parents

**Comment by Bessonov** [comment]
Run in the same challenge. I have an entity, let say A, with a little bit complex permission rules. Entity B has a FK to entity A and entity C has a FK to entity B. Both B and C follows the same rules like entity A: is entity A selectable, then B&C are selectable too. Same for update and delete. It would be great if I could specify in entity C something like `{"b":{"a":{"_use_perm_from": "select"}}}`.

**Comment by altschuler** [comment]
> > 1. Delegation by foreign key: on a group_post (a table containing posts in a group), with a foreign key to groups, you would say that if you have access to the referenced group, you also have access to the group_post.
> 
> This is probably a valid approach since if the referenced table row's key is not accessible because of some permissions seems logical that another table rows referencing it through foreign key shouldn't also be accessible. The problem is a possible explosion of joins when the algorithm is gonna traverse upwards and for each table does the same thing -- check if the rows having foregin keys are accessible through parents

Valid point, but I don't think it would be much more of an explosion than creating those rules manually all the way up the hierarchy?

**Comment by nikhilshinday** [comment]
hey @altschuler! I found a workaround that you may find beneficial. 

If you think of the permissions DSL that hasura exposes as, effectively, a SQL query expressed in a graph-like DSL, you can create permissions definitions in views in the database, and then use those views wherever you want to delegate permissions. 

## The old way of doing things
As an example, let's say that you have a session schema for a user: 
{x-hasura-user-id: bigint, x-hasura-organization-id: bigint}

And a schema shape: 
```
table Organization {
  id: bigint pk;
```

```
table User {
  id: bigint pk;
  organization_id: bigint fk;
}
```

```
table Parent {
  id: bigint pk;
  creator_id: bigint fk;
  organization_id: bigint fk;
  public: boolean;
```

```
table Child {
  id: bigint pk;
  parent_id: bigint fk;
  content: text
}
```
with a relationship defined from Child -> Parent named "parent"

The way that we would typically define select permissions on the parent is:

```
_or: {
  { _and: [ { public: { _eq: false } }, { creator_id: { _eq: "x-hasura-user-id" } }] },
  { _and: [ { public: { _eq: true } }, { organization_id: { _eq: "x-hasura-organization-id" } }] }
}
```

and the way that we would typically define the select permission on the child is:
```
parent: {
  _or: {
    { _and: [ { public: { _eq: false } }, { creator_id: { _eq: "x-hasura-user-id" } }] },
    { _and: [ { public: { _eq: true } }, { organization_id: { _eq: "x-hasura-organization-id" } }] }
  }
}
```

The problem here is that the query can get complex very quickly, and every time that it does, you will need to update every single child. For a sufficiently important parent table, you can have lots of children!

## The hack that I found

Instead of defining the select permission in the permissions DSL, create a new view:
```
create or replace view parent_user_select_permission as (
  select creator_id user_id, id parent_id from parent
) union distinct (
  select user.id user_id, parent.id parent_id from parent 
    join user on parent.organization_id = user.organization_id
    where parent.public = true
);
```

and then create an array relationship between the view and the parent (defined below using the tables.yaml format): 
```
- table:
    name: parent
    schema: public
  array_relationships:
    - name: select_permission
      using:
        manual_configuration:
          column_mapping:
            id: parent_id
          insertion_order: null
          remote_table:
            name: parent_user_select_permission
            schema: public
```

Define the parent permission as: 
```
{ select_permission: { user_id: { _eq: "x-hasura-user-id" } } }
```
and the child permission as: 
```
{ select_permission: { parent_id: { _eq: "x-hasura-user-id } } }
```

The main benefits of this are: 
1. you can create complex permissions statements without having to repeat yourself for delegated models
2. you can swap out the underlying statement without needing to change permissions in Hasura as long as the underlying structure of delegation is unchanged
3. permissions in the hasura dashboard are more concise 

The downside, as with any attempt at DRY, is that you create a layer of indirection when sometimes you may not need to have it.

My general solution to avoiding that downside is to only do it for really complicated permissions expressions OR when I have to delegate permissions to a ton of models. 

Hope you found that helpful!

**Comment by Bessonov** [comment]
@nikhilshinday The issue is that it becomes very messy, and you lose all the readability benefits.

**Comment by nikhilshinday** [comment]
@Bessonov you're absolutely right: the tradeoff for readability is flexibility and abstraction. I've built an entire B2B app off of Hasura, and having the ability to abstract away permissions into views offers reusability in my backend in addition to being able to escape the DSL's limitations. It also prolongs the time before which I'll have to replace Hasura with a home-grown solution, which will easily take 6+ months of developer time. I consider those benefits to be favorable enough to accept the readability tradeoff, but YMMV based on your own use case 😃
